File: 2816.double-a-number-represented-as-a-linked-list.java
#
# @lc app=leetcode id=2816 lang=java
#
# [2816] Double a Number Represented as a Linked List
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode doubleIt(ListNode head) {
        ListNode revHead = reverseList(head);
        ListNode curr = revHead;
        int carry = 0;
        ListNode tail = null;
        while (curr != null) {
            int prod = curr.val * 2 + carry;
            curr.val = prod % 10;
            carry = prod / 10;
            tail = curr;
            curr = curr.next;
        }
        if (carry != 0) {
            tail.next = new ListNode(carry);
        }
        return reverseList(revHead);
    }
    
    private ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode curr = head;
        while (curr != null) {
            ListNode nextTemp = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextTemp;
        }
        return prev;
    }
}
# @lc code=end

File: 2816.double-a-number-represented-as-a-linked-list.go
#
# @lc app=leetcode id=2816 lang=golang
#
# [2816] Double a Number Represented as a Linked List
#
# @lc code=start
/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/
func doubleIt(head *ListNode) *ListNode {
    carry := helper(head)
    if carry > 0 {
        newHead := &ListNode{Val: carry, Next: head}
        return newHead
    }
    return head
}

func helper(node *ListNode) int {
    if node == nil {
        return 0
    }
    
    // Process the rest of the list first
    carry := helper(node.Next)
    
    // Double current digit and add carry from right
    doubled := node.Val * 2 + carry
    node.Val = doubled % 10
    return doubled / 10
}
# @lc code=end